### ISSUE TITLE
DTT1 - Iteration 3 - WorkFlow engine module - Perform batery of tests

### ISSUE THREAD

**Issue by fcaffieri** [wazuh/wazuh-qa#4910]
> Epic: https://github.com/wazuh/wazuh-qa/issues/4905

---

## Description

The objective of this issue is to carry out a battery of tests for the end user, in order to detect possible failures and ensure the correct functionality of the module to be released.

## Taks

- [x] Generate use cases.
- [x] Perform tests on the generated use cases.

## Test Cases

[Evidence test 1 to 32](https://github.com/wazuh/wazuh-qa/issues/4910#issuecomment-1946352275)
[Evidence test 33 to 63](https://github.com/wazuh/wazuh-qa/issues/4910#issuecomment-1947431407)

| Test Case | Host | Test | Precondition | Expected Result | Result obtained |
| :---: | :---: | :---: | :---: | :---: | :---: |
1|Agent & Manager|Happy Path all modules|general data ok| general data ok| :green_circle:
|||allocation data ok|allocation ok
|||provision data ok|provision ok
|||testing data ok|testing ok
2|Agent & Manager|General parameters variables-agent error |general data ok, wrong description-agent|Failure|:green_circle:
3|Agent & Manager|General parameters variables-manager error |general data ok, wrong description-manager|Failure|:green_circle:
4|Agent & Manager|General parameters variables-infra-provider error |general data ok, wrong infra-provider|Failure|:green_circle:
5|Agent|Allocation testing wrong do-this|general data ok|Failure|:green_circle:
|||general data ok|allocation data with wrong do-this value
6|Agent|Allocation testing wrong with-path|general data ok|Failure|:green_circle:
|||allocation data with wrong with-path
7|Agent|Allocation testing wrong with-args-main-path|general data ok|Failure|:green_circle:
|||allocation data with wrong with-args-main-path
8|Agent|Allocation testing wrong with-args-action|general data ok|Failure|:green_circle:
|||allocation data with wrong with-args-action
9|Agent|Allocation testing wrong with-args-provider|general data ok|Failure|:green_circle:
|||allocation data with wrong with-args-provider
10|Agent|Allocation testing wrong with-args-size|general data ok|Failure|:green_circle:
|||allocation data with wrong with-args-size
11|Agent|Allocation testing wrong with-args-composite-name|general data ok|Failure|:green_circle:
|||allocation data with wrong with-args-composite-name
12|Agent|Allocation testing wrong with-args-inventory-output|general data ok|Failure|:green_circle:
|||allocation data with wrong with-args-inventory-output
13|Agent|Allocation testing wrong with-args-track-output|general data ok|Failure|:green_circle:
|||allocation data with wrong with-args-track-output
14|Manager|Allocation testing wrong do-this|general data ok|Failure|:green_circle:
|||allocation data with wrong do-this value
15|Manager|Allocation testing wrong with-path|general data ok|Failure|:green_circle:
|||allocation data with wrong with-path
16|Manager|Allocation testing wrong with-args-main-path|general data ok|Failure|:green_circle:
|||allocation data with wrong with-args-main-path
17|Manager|Allocation testing wrong with-args-action|general data ok|Failure|:green_circle:
|||allocation data with wrong with-args-action
18|Manager|Allocation testing wrong with-args-provider|general data ok|Failure|:green_circle:
|||allocation data with wrong with-args-provider
19|Manager|Allocation testing wrong with-args-size|general data ok|Failure|:green_circle:
|||allocation data with wrong with-args-size
20|Manager|Allocation testing wrong with-args-composite-name|general data ok|Failure|:green_circle:
|||allocation data with wrong with-args-composite-name
21|Manager|Allocation testing wrong with-args-inventory-output|general data ok|Failure|:green_circle:
|||allocation data with wrong with-args-inventory-output
22|Manager|Allocation testing wrong with-args-track-output|general data ok|Failure|:green_circle:
|||allocation data with wrong with-args-track-output
23|Agent or Manager|Allocation testing cleanup|general data ok|Clean up after provisioning|:green_circle:
|||allocation data with cleanup data ok
24|Agent or Manager|Allocation testing cleanup wrong with-path|general data ok|clean up failure|:green_circle:
|||allocation data with wrong with-path
25|Agent or Manager|Allocation testing cleanup wrong with-args-main|general data ok|clean up failure|:green_circle:
|||allocation data with wrong with-args-main
26|Agent or Manager|Allocation testing cleanup wrong with-args-action|general data ok|clean up failure|:green_circle:
|||allocation data with wrong with-args-action
27|Agent or Manager|Allocation testing cleanup wrong with-args-track-output|general data ok|clean up failure|:green_circle:
|||allocation data with wrong with-args-track-output
28|Agent|Provision wrong do-this|general data ok|Failure|:green_circle:
|||allocation data ok
|||provision data with wrong do-this
29|Agent|Provision wrong with-path|general data ok|Failure|:green_circle:
|||allocation data ok
|||provision data with wrong with-path
30|Agent|Provision wrong with-args-inventory-agent|general data ok|Failure|:green_circle:
|||allocation data ok
|||provision data with wrong with-args-inventory-agent
31|Agent|Provision wrong with-args-inventory-manager|general data ok|Failure|:green_circle:
|||allocation data ok
|||provision data with wrong with-args-inventory-manager
32|Agent|Provision wrong with-args-install-component|general data ok|Failure|:green_circle:
|||allocation data ok
|||provision data with wrong with-args-install-component
33|Agent|Provision wrong with-args-install-type|general data ok|Failure|:green_circle:
|||allocation data ok
|||provision data with wrong with-args-install-type
34|Agent|Provision wrong with-args-install-component|general data ok|Failure|:yellow_circle:
|||allocation data ok
|||provision data with wrong with-args-install-component
35|Manager|Provision wrong do-this|general data ok|Failure|:green_circle:
|||allocation data ok
|||provision data with wrong do-this
36|Manager|Provision wrong with-path|general data ok|Failure|:green_circle:
|||allocation data ok
|||provision data with wrong with-path
37|Manager|Provision wrong with-args-inventory-agent|general data ok|Failure|:green_circle:
|||allocation data ok
|||provision data with wrong with-args-inventory-agent
38|Manager|Provision wrong with-args-inventory-manager|general data ok|Failure|:green_circle:
|||allocation data ok
|||provision data with wrong with-args-inventory-manager
39|Manager|Provision wrong with-args-install-component|general data ok|Failure|:yellow_circle:
|||allocation data ok
|||provision data with wrong with-args-install-component
40|Manager|Provision wrong with-args-install-type|general data ok|Failure|:green_circle:
|||allocation data ok
|||provision data with wrong with-args-install-type
41|Manager|Provision wrong with-args-install-component|general data ok|Failure|:green_circle:
|||allocation data ok
|||provision data with wrong with-args-install-component
42|Agent or Manager|Provision foreach|general data ok|provision to all the declared hosts only|:green_circle:
|||allocation data ok
|||provision data with for each ok
43|Agent or Manager|Provision foreach wrong foreach-variable|general data ok|Failure|:green_circle:
|||allocation data ok
|||provision data with wrong foreach-variable
44|Agent or Manager|Provision foreach wrong foreach-as|general data ok|Failure|:green_circle:
|||allocation data ok
|||provision data with wrong foreach-as
45|Agent or Manager|Provision depends-on|general data ok|Provision after the previous modules|:green_circle:
|||allocation data ok
|||provision data with depends-on on previous modules
46|Agent or Manager|Provision depends-on wrong data|general data ok|Failure|:green_circle:
|||allocation data ok
|||provision data with wrong modules
47|Agent|Testing with wrong do-this|general data ok|Failure|:green_circle:
|||allocation data ok
|||provision data ok
|||testing data with wrong do-this
48|Agent|Testing with wrong with-path|general data ok|Failure|:green_circle:
|||allocation data ok
|||provision data ok
|||testing data with wrong with-path
49|Agent|Testing with wrong with-args-main-path|general data ok|Failure|:green_circle:
|||allocation data ok
|||provision data ok
|||testing data with wrong  with-args-main
50|Agent|Testing with wrong with-args-inventory|general data ok|Failure|:green_circle:
|||allocation data ok
|||provision data ok
|||testing data with wrong with-args-inventory
51|Agent|Testing with wrong with-args-dependencies|general data ok|Failure|:green_circle:
|||allocation data ok
|||provision data ok
|||testing data with wrong with-args-dependencies
52|Agent|Testing with wrong with-args-tests |general data ok|Failure|:green_circle:
|||allocation data ok
|||provision data ok
|||testing data with wrong with-args-tests 
53|Manager|Testing with wrong do-this|general data ok|Failure|:green_circle:
|||allocation data ok
|||provision data ok
|||testing data with wrong do-this
54|Manager|Testing with wrong with-path|general data ok|Failure|:green_circle:
|||allocation data ok
|||provision data ok
|||testing data with wrong with-path
55|Manager|Testing with wrong with-args-main-path|general data ok|Failure|:green_circle:
|||allocation data ok
|||provision data ok
|||testing data with wrong  with-args-main
56|Manager|Testing with wrong with-args-inventory|general data ok|Failure|:green_circle:
|||allocation data ok
|||provision data ok
|||testing data with wrong with-args-inventory
57|Manager|Testing with wrong with-args-dependencies|general data ok|Failure|:green_circle:
|||allocation data ok
|||provision data ok
|||testing data with wrong with-args-dependencies
58|Manager|Testing with wrong with-args-tests |general data ok|Failure|:green_circle:
|||allocation data ok
|||provision data ok
|||testing data with wrong with-args-tests 
59|Agent or Manager|Testing foreach|general data ok|Tests to all the declared hosts only|:green_circle:
|||allocation data ok
|||provision data ok
|||testing data with for each ok
60|Agent or Manager|Testing foreach wrong foreach-variable|general data ok|Failure|:green_circle:
|||allocation data ok
|||provision data ok
|||testing data with wrong foreach-variable
61|Agent or Manager|Testing foreach wrong foreach-variable|general data ok|Failure|:green_circle:
|||allocation data ok
|||provision data ok
|||testing data with wrong foreach-as
62|Agent or Manager|Testing depends-on|general data ok|Tests after the previous modules|:green_circle:
|||allocation data ok
|||provision data ok
|||testing data  with depends-on on previous modules
63|Agent or Manager|Testing depends-on wrong data|general data ok|Failure|:green_circle:
|||allocation data ok
|||provision data ok
|||provision data with wrong modules

### Modules test 
> Note: these tests are not a separated test. It is a checklist based in Workflow tests

|Test Number|	Host|	Test|	Precondition|	Checklist| Result |
|---|---|---|---|---|---|
|1|	Agent & Manager|	Check allocation|	general data ok|	
| |                |      allocation data ok|	Check hosts were created as fixture (size and OS) | 
| |                |                        ||Check if in the working-dir defined in data each host has a directory |:green_circle:
| |                |                        ||Check the content inventory.yaml (ansible_host, ansible_port, ansible_ssh_private_key, ansible_user) | :green_circle:
| |                |                        ||Check the content track.yaml (identifier, instance_dir, keypath, provider) | :green_circle:
| |                |                        ||Check the ips, users, ports and private_key| :green_circle:
|2|	Agent & Manager|	Check provision|	general data ok|	
| |                |      allocation data ok|	
| |                |      provision data ok||Check the installation/uninstallation of the application declared in the fixture | :green_circle:
| |                |      provision data ok||Check the status of  the application declared in the fixture |:red_circle:
| |                |                        ||If it is Wazuh, check that the type of installation done is correct ansible_user) | :green_circle:
|3|	Agent & Manager|	Check testing|	general data ok|	
| |                |      allocation data ok|	
| |                |      provision data ok||Check the results of the test | :green_circle:
| |                |     testing data ok ||Run the tests manually if is possible to confirm | :green_circle:
| |                |                        ||If Wazuh is tested, check the version and revision of tested Wazuh is correct | :green_circle:
| |                |                        ||If it is Wazuh manager is connected to the agent| :red_circle:
|4|	Agent & Manager|	Check Worfklow logs|	general data no/ok|	Check error handling and logs | :green_circle:
| |                |      allocation data no/ok|	
| |                |      provision data no/ok||
| |                |     testing data no/ok ||

**Comment by pro-akim** [comment]
## Update
I have started developing the test cases, however, I will continue after finish https://github.com/wazuh/wazuh-qa/issues/4909
Moved to on hold

**Comment by pro-akim** [comment]
## Update

Retaking the issue after weekly security report is done.

**Comment by pro-akim** [comment]
## Update

Some tests are meant to check the proper functioning of the workflow development flow according to the fixture, referred to as `Testing Workflow flows `Other tests are more for verification and are called `Testing modules`.

[test_cases_wf.xls](https://github.com/wazuh/wazuh-qa/files/14212713/test_cases_wf.xls)

**Comment by pro-akim** [comment]
## Update
Testing...

**Comment by fcaffieri** [comment]
### Update

### Test (WIP)


<details> <summary> Test </summary> 

Test number 63: :green_circle: 

```
  - task: "provision-manager"
    description: "Provision the manager."
    do:
      this: process
      with:
        path: python3
        args:
          - provision/main.py
          - inventory-manager: "{working-dir}/manager-{manager-os}/inventory.yaml"
          - install:
            - component: wazuh-manager
              type: package
    depends-on:
      - "allocate-manager-not-exist"
```

Obtained result:

```

$ python3 -m workflow_engine workflow_engine/examples/dtt1-agents-poc.yaml
Error while provisioning: Tasks do not exist: allocate-manager-not-exist
$ 

```

Test number 62: :green_circle: 

```

# Copyright (C) 2015, Wazuh Inc.
# Created by Wazuh, Inc. <info@wazuh.com>.
# This program is a free software; you can redistribute it and/or modify it under the terms of GPLv2
version: 0.1
description: This workflow is used to test agents deployment por DDT1 PoC
variables:
  agents-os:
    - linux-ubuntu-22.04-amd64
  manager-os: linux-ubuntu-22.04-amd64
  infra-provider: vagrant
  working-dir: /tmp/dtt1-poc

tasks:
  # Generic agent test task
  - task: "run-agent-tests-{agent}"
    description: "Run tests uninstall for the {agent} agent."
    do:
      this: process
      with:
        path: python3
        args:
          - testing/main.py
          - inventory: "{working-dir}/agent-{agent}/inventory.yaml"
          - dependencies:
            - manager: "{working-dir}/manager-{manager-os}/inventory.yaml"
            - agent: "{working-dir}/agent-{agent}/inventory.yaml"
          - tests: "install,register,stop"
          - component: "agent"
          - wazuh-version: "4.7.1"
          - wazuh-revision: "40709"
    depends-on:
      - "provision-install-{agent}"
      - "provision-manager"
    foreach:
      - variable: agents-os
        as: agent

  # Generic agent test task
  - task: "run-agent-tests-uninstall-{agent}"
    description: "Run tests uninstall for the {agent} agent."
    do:
      this: process
      with:
        path: python3
        args:
          - testing/main.py
          - inventory: "{working-dir}/agent-{agent}/inventory.yaml"
          - dependencies:
            - manager: "{working-dir}/manager-{manager-os}/inventory.yaml"
          - tests: "uninstall"
          - component: "agent"
          - wazuh-version: "4.7.1"
          - wazuh-revision: "40709"
    depends-on:
      - "run-agent-tests-{agent}"
      - "provision-uninstall-{agent}"
    foreach:
      - variable: agents-os
        as: agent

  # Unique manager provision task
  - task: "provision-manager"
    description: "Provision the manager."
    do:
      this: process
      with:
        path: python3
        args:
          - provision/main.py
          - inventory-manager: "{working-dir}/manager-{manager-os}/inventory.yaml"
          - install:
            - component: wazuh-manager
              type: package
    depends-on:
      - "allocate-manager"

  # Unique manager allocate task
  - task: "allocate-manager"
    description: "Allocate resources for the manager."
    do:
      this: process
      with:
        path: python3
        args:
          - allocation/main.py
          - action: create
          - provider: "{infra-provider}"
          - size: large
          - composite-name: "{manager-os}"
          - inventory-output: "{working-dir}/manager-{manager-os}/inventory.yaml"
          - track-output: "{working-dir}/manager-{manager-os}/track.yaml"
    cleanup:
      this: process
      with:
        path: python3
        args:
          - allocation/main.py
          - action: delete
          - track-output: "{working-dir}/manager-{manager-os}/track.yaml"

  # Generic agent provision task
  - task: "provision-install-{agent}"
    description: "Provision resources for the {agent} agent."
    do:
      this: process
      with:
        path: python3
        args:
          - provision/main.py
          - inventory-agent: "{working-dir}/agent-{agent}/inventory.yaml"
          - inventory-manager: "{working-dir}/manager-{manager-os}/inventory.yaml"
          - install:
            - component: wazuh-agent
              type: package
            - component: curl
    depends-on:
      - "allocate-{agent}"
      - "provision-manager"
    foreach:
      - variable: agents-os
        as: agent

  # Generic agent provision task
  - task: "provision-uninstall-{agent}"
    description: "Provision resources for the {agent} agent."
    do:
      this: process
      with:
        path: python3
        args:
          - provision/main.py
          - inventory-agent: "{working-dir}/agent-{agent}/inventory.yaml"
          - inventory-manager: "{working-dir}/manager-{manager-os}/inventory.yaml"
          - uninstall:
            - component: wazuh-agent
              type: package
    depends-on:
      - "provision-install-{agent}"
    foreach:
      - variable: agents-os
        as: agent

  # Generic agent allocate task
  - task: "allocate-{agent}"
    description: "Allocate resources for the {agent} agent."
    do:
      this: process
      with:
        path: python3
        args:
          - allocation/main.py
          - action: create
          - provider: "{infra-provider}"
          - size: small
          - composite-name: "{agent}"
          - inventory-output: "{working-dir}/agent-{agent}/inventory.yaml"
          - track-output: "{working-dir}/agent-{agent}/track.yaml"
    cleanup:
      this: process
      with:
        path: python3
        args:
          - allocation/main.py
          - action: delete
          - track-output: "{working-dir}/agent-{agent}/track.yaml"
    foreach:
      - variable: agents-os
        as: agent

```

Obtained result:

```
$ python3 -m workflow_engine workflow_engine/examples/dtt1-agents-poc.yaml
[2024-02-15 18:02:05] [INFO] [444622] [MainThread] [workflow_engine]: Executing DAG tasks.
[2024-02-15 18:02:05] [INFO] [444622] [MainThread] [workflow_engine]: Executing tasks in parallel.
[2024-02-15 18:02:05] [INFO] [444622] [ThreadPoolExecutor-0_0] [workflow_engine]: [allocate-manager] Starting task.
[2024-02-15 18:02:50] [INFO] [444622] [ThreadPoolExecutor-0_0] [workflow_engine]: [allocate-manager] Finished task in 45.12 seconds.
[2024-02-15 18:02:50] [INFO] [444622] [ThreadPoolExecutor-0_0] [workflow_engine]: [allocate-linux-ubuntu-22.04-amd64] Starting task.
[2024-02-15 18:03:39] [INFO] [444622] [ThreadPoolExecutor-0_0] [workflow_engine]: [allocate-linux-ubuntu-22.04-amd64] Finished task in 49.15 seconds.
[2024-02-15 18:03:39] [INFO] [444622] [ThreadPoolExecutor-0_0] [workflow_engine]: [provision-manager] Starting task.
[2024-02-15 18:05:44] [INFO] [444622] [ThreadPoolExecutor-0_0] [workflow_engine]: [provision-manager] Finished task in 124.63 seconds.
[2024-02-15 18:05:44] [INFO] [444622] [ThreadPoolExecutor-0_0] [workflow_engine]: [provision-install-linux-ubuntu-22.04-amd64] Starting task.
[2024-02-15 18:06:55] [INFO] [444622] [ThreadPoolExecutor-0_0] [workflow_engine]: [provision-install-linux-ubuntu-22.04-amd64] Finished task in 71.12 seconds.
[2024-02-15 18:06:55] [INFO] [444622] [ThreadPoolExecutor-0_0] [workflow_engine]: [run-agent-tests-linux-ubuntu-22.04-amd64] Starting task.
[2024-02-15 18:07:17] [INFO] [444622] [ThreadPoolExecutor-0_0] [workflow_engine]: [run-agent-tests-linux-ubuntu-22.04-amd64] Finished task in 21.68 seconds.
[2024-02-15 18:07:17] [INFO] [444622] [ThreadPoolExecutor-0_0] [workflow_engine]: [provision-uninstall-linux-ubuntu-22.04-amd64] Starting task.
[2024-02-15 18:07:31] [INFO] [444622] [ThreadPoolExecutor-0_0] [workflow_engine]: [provision-uninstall-linux-ubuntu-22.04-amd64] Finished task in 13.81 seconds.
[2024-02-15 18:07:31] [INFO] [444622] [ThreadPoolExecutor-0_0] [workflow_engine]: [run-agent-tests-uninstall-linux-ubuntu-22.04-amd64] Starting task.
[2024-02-15 18:07:46] [INFO] [444622] [ThreadPoolExecutor-0_0] [workflow_engine]: [run-agent-tests-uninstall-linux-ubuntu-22.04-amd64] Finished task in 15.06 seconds.
[2024-02-15 18:07:46] [INFO] [444622] [MainThread] [workflow_engine]: Executing Reverse DAG tasks.
[2024-02-15 18:07:46] [INFO] [444622] [MainThread] [workflow_engine]: Executing tasks in parallel.
[2024-02-15 18:07:46] [INFO] [444622] [ThreadPoolExecutor-1_0] [workflow_engine]: [allocate-linux-ubuntu-22.04-amd64] Starting task.
[2024-02-15 18:07:46] [INFO] [444622] [ThreadPoolExecutor-0_0] [workflow_engine]: [allocate-linux-ubuntu-22.04-amd64] Finished task in 37.42 seconds.
[2024-02-15 18:07:50] [INFO] [444622] [ThreadPoolExecutor-0_0] [workflow_engine]: [allocate-manager] Starting task.
[2024-02-15 18:07:50] [INFO] [444622] [ThreadPoolExecutor-0_0] [workflow_engine]: [allocate-manager] Finished task in 40.41 seconds.
[2024-02-15 18:07:51] [INFO] [141693] [MainThread] [workflow_engine]: Executing Reverse DAG tasks.
[2024-02-15 18:07:51] [INFO] [141693] [MainThread] [workflow_engine]: Executing tasks in parallel.
```


Test number 61: :green_circle: 

```
- task: "run-agent-tests-{agent}"
    description: "Run tests uninstall for the {agent} agent."
    do:
      this: process
      with:
        path: python3
        args:
          - testing/main.py
          - inventory: "{working-dir}/agent-{agent}/inventory.yaml"
          - dependencies:
            - manager: "{working-dir}/manager-{manager-os}/inventory.yaml"
            - agent: "{working-dir}/agent-{agent}/inventory.yaml"
          - tests: "install,register,stop"
          - component: "agent"
          - wazuh-version: "4.7.1"
          - wazuh-revision: "40709"
    #depends-on:
    #  - "provision-install-{agent}"
    #  - "provision-manager"
    foreach:
      - variable: agents-os
        as: not-exist
```

Obtained result:

```
$ python3 -m workflow_engine workflow_engine/examples/dtt1-agents-poc.yaml 
Error while provisioning: 'agent'
$
```

Test number 60: :green_circle: 

```
  - task: "run-agent-tests-{agent}"
    description: "Run tests uninstall for the {agent} agent."
    do:
      this: process
      with:
        path: python3
        args:
          - testing/main.py
          - inventory: "{working-dir}/agent-{agent}/inventory.yaml"
          - dependencies:
            - manager: "{working-dir}/manager-{manager-os}/inventory.yaml"
            - agent: "{working-dir}/agent-{agent}/inventory.yaml"
          - tests: "install,register,stop"
          - component: "agent"
          - wazuh-version: "4.7.1"
          - wazuh-revision: "40709"
    #depends-on:
    #  - "provision-install-{agent}"
    #  - "provision-manager"
    foreach:
      - variable: not-exist
        as: agent
```

Obtained result:

```
$ python3 -m workflow_engine workflow_engine/examples/dtt1-agents-poc.yaml 
Error while provisioning: Tasks do not exist: run-agent-tests-linux-ubuntu-22.04-amd64
$
```


Test number 59: :green_circle: 

```
# Copyright (C) 2015, Wazuh Inc.
# Created by Wazuh, Inc. <info@wazuh.com>.
# This program is a free software; you can redistribute it and/or modify it under the terms of GPLv2
version: 0.1
description: This workflow is used to test agents deployment por DDT1 PoC
variables:
  agents-os:
    - linux-ubuntu-22.04-amd64
  manager-os: linux-ubuntu-22.04-amd64
  infra-provider: vagrant
  working-dir: /tmp/dtt1-poc

tasks:
  # Generic agent test task
  - task: "run-agent-tests-{agent}"
    description: "Run tests uninstall for the {agent} agent."
    do:
      this: process
      with:
        path: python3
        args:
          - testing/main.py
          - inventory: "{working-dir}/agent-{agent}/inventory.yaml"
          - dependencies:
            - manager: "{working-dir}/manager-{manager-os}/inventory.yaml"
            - agent: "{working-dir}/agent-{agent}/inventory.yaml"
          - tests: "install,register,stop"
          - component: "agent"
          - wazuh-version: "4.7.1"
          - wazuh-revision: "40709"
    depends-on:
      - "provision-install-{agent}"
      - "provision-manager"
    foreach:
      - variable: agents-os
        as: agent

  

  # Unique manager provision task
  - task: "provision-manager"
    description: "Provision the manager."
    do:
      this: process
      with:
        path: python3
        args:
          - provision/main.py
          - inventory-manager: "{working-dir}/manager-{manager-os}/inventory.yaml"
          - install:
            - component: wazuh-manager
              type: package
    depends-on:
      - "allocate-manager"

  # Unique manager allocate task
  - task: "allocate-manager"
    description: "Allocate resources for the manager."
    do:
      this: process
      with:
        path: python3
        args:
          - allocation/main.py
          - action: create
          - provider: "{infra-provider}"
          - size: large
          - composite-name: "{manager-os}"
          - inventory-output: "{working-dir}/manager-{manager-os}/inventory.yaml"
          - track-output: "{working-dir}/manager-{manager-os}/track.yaml"
    cleanup:
      this: process
      with:
        path: python3
        args:
          - allocation/main.py
          - action: delete
          - track-output: "{working-dir}/manager-{manager-os}/track.yaml"

  # Generic agent provision task
  - task: "provision-install-{agent}"
    description: "Provision resources for the {agent} agent."
    do:
      this: process
      with:
        path: python3
        args:
          - provision/main.py
          - inventory-agent: "{working-dir}/agent-{agent}/inventory.yaml"
          - inventory-manager: "{working-dir}/manager-{manager-os}/inventory.yaml"
          - install:
            - component: wazuh-agent
              type: package
            - component: curl
    depends-on:
      - "allocate-{agent}"
      - "provision-manager"
    foreach:
      - variable: agents-os
        as: agent

 

  # Generic agent allocate task
  - task: "allocate-{agent}"
    description: "Allocate resources for the {agent} agent."
    do:
      this: process
      with:
        path: python3
        args:
          - allocation/main.py
          - action: create
          - provider: "{infra-provider}"
          - size: small
          - composite-name: "{agent}"
          - inventory-output: "{working-dir}/agent-{agent}/inventory.yaml"
          - track-output: "{working-dir}/agent-{agent}/track.yaml"
    cleanup:
      this: process
      with:
        path: python3
        args:
          - allocation/main.py
          - action: delete
          - track-output: "{working-dir}/agent-{agent}/track.yaml"
    foreach:
      - variable: agents-os
        as: agent
```

Obtained result:

```
$ python3 -m workflow_engine workflow_engine/examples/dtt1-agents-poc.yaml 
[2024-02-15 18:21:35] [INFO] [454631] [MainThread] [workflow_engine]: Executing DAG tasks.
[2024-02-15 18:21:35] [INFO] [454631] [MainThread] [workflow_engine]: Executing tasks in parallel.
[2024-02-15 18:21:35] [INFO] [454631] [ThreadPoolExecutor-0_0] [workflow_engine]: [allocate-manager] Starting task.
[2024-02-15 18:22:19] [INFO] [454631] [ThreadPoolExecutor-0_0] [workflow_engine]: [allocate-manager] Finished task in 43.67 seconds.
[2024-02-15 18:22:19] [INFO] [454631] [ThreadPoolExecutor-0_0] [workflow_engine]: [allocate-linux-ubuntu-22.04-amd64] Starting task.
[2024-02-15 18:23:07] [INFO] [454631] [ThreadPoolExecutor-0_0] [workflow_engine]: [allocate-linux-ubuntu-22.04-amd64] Finished task in 48.46 seconds.
[2024-02-15 18:23:07] [INFO] [454631] [ThreadPoolExecutor-0_0] [workflow_engine]: [provision-manager] Starting task.
[2024-02-15 18:25:15] [INFO] [454631] [ThreadPoolExecutor-0_0] [workflow_engine]: [provision-manager] Finished task in 127.61 seconds.
[2024-02-15 18:25:15] [INFO] [454631] [ThreadPoolExecutor-0_0] [workflow_engine]: [provision-install-linux-ubuntu-22.04-amd64] Starting task.
[2024-02-15 18:26:40] [INFO] [454631] [ThreadPoolExecutor-0_0] [workflow_engine]: [provision-install-linux-ubuntu-22.04-amd64] Finished task in 85.00 seconds.
[2024-02-15 18:26:40] [INFO] [454631] [ThreadPoolExecutor-0_0] [workflow_engine]: [run-agent-tests-linux-ubuntu-22.04-amd64] Starting task.
[2024-02-15 18:27:06] [INFO] [454631] [ThreadPoolExecutor-0_0] [workflow_engine]: [run-agent-tests-linux-ubuntu-22.04-amd64] Finished task in 26.17 seconds.
[2024-02-15 18:27:06] [INFO] [454631] [MainThread] [workflow_engine]: Executing Reverse DAG tasks.
[2024-02-15 18:27:06] [INFO] [454631] [MainThread] [workflow_engine]: Executing tasks in parallel.
[2024-02-15 18:27:06] [INFO] [454631] [ThreadPoolExecutor-1_0] [workflow_engine]: [allocate-linux-ubuntu-22.04-amd64] Starting task.
[2024-02-15 18:27:12] [INFO] [454631] [ThreadPoolExecutor-1_0] [workflow_engine]: [allocate-linux-ubuntu-22.04-amd64] Finished task in 6.23 seconds.
[2024-02-15 18:27:12] [INFO] [454631] [ThreadPoolExecutor-1_0] [workflow_engine]: [allocate-manager] Starting task.
[2024-02-15 18:27:18] [INFO] [454631] [ThreadPoolExecutor-1_0] [workflow_engine]: [allocate-manager] Finished task in 6.23 seconds.
$ 
```


Test number 58: :green_circle: 

```
tasks:
  # Generic manager test task
  - task: "run-manager-tests"
    description: "Run tests uninstall for the manager."
    do:
      this: process
      with:
        path: python3
        args:
          - testing/main.py
          - inventory: "{working-dir}/manager/inventory.yaml"
          - tests: "upgrade"
          - component: "manager"
          - wazuh-version: "4.7.1"
          - wazuh-revision: "40709"
    depends-on:
      #- "provision-install-{manager}"
      - "provision-manager"

  # Unique manager provision task
  - task: "provision-manager"
    description: "Provision the manager."
    do:
      this: process
      with:
        path: python3
        args:
          - provision/main.py
          - inventory-manager: "{working-dir}/manager-{manager-os}/inventory.yaml"
          - install:
            - component: wazuh-manager
              type: package
    depends-on:
      - "allocate-manager"

  # Unique manager allocate task
  - task: "allocate-manager"
    description: "Allocate resources for the manager."
    do:
      this: process
      with:
        path: python3
        args:
          - allocation/main.py
          - action: create
          - provider: "{infra-provider}"
          - size: large
          - composite-name: "{manager-os}"
          - inventory-output: "{working-dir}/manager-{manager-os}/inventory.yaml"
          - track-output: "{working-dir}/manager-{manager-os}/track.yaml"
    cleanup:
      this: process
      with:
        path: python3
        args:
          - allocation/main.py
          - action: delete
          - track-output: "{working-dir}/manager-{manager-os}/tr(.venv) fcaffieri@Fede:~/repos/wazuh-qa/deployability/moduleslow_engine]: Executing tasks in parallel.
[2024-02-15 18:52:02] [INFO] [465591] [ThreadPoolExecutor-0_0] [workflow_engine]: [allocate-manager] Starting task.
[2024-02-15 18:52:43] [INFO] [465591] [ThreadPoolExecutor-0_0] [workflow_engine]: [allocate-manager] Finished task in 41.01 seconds.
[2024-02-15 18:52:43] [INFO] [465591] [ThreadPoolExecutor-0_0] [workflow_engine]: [provision-manager] Starting task.
[2024-02-15 18:55:03] [INFO] [465591] [ThreadPoolExecutor-0_0] [workflow_engine]: [provision-manager] Finished task in 139.38 seconds.
[2024-02-15 18:55:03] [INFO] [465591] [ThreadPoolExecutor-0_0] [workflow_engine]: [run-manager-tests] Starting task.
[2024-02-15 18:55:03] [ERROR] [465591] [ThreadPoolExecutor-0_0] [workflow_engine]: [run-manager-tests] Task failed with error: Error executing process task Traceback (most recent call last):
  File "/home/fcaffieri/repos/wazuh-qa/deployability/modules/testing/main.py", line 23, in <module>
    Tester.run(InputPayload(**vars(parse_arguments())))
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/fcaffieri/repos/wazuh-qa/.venv/lib/python3.11/site-packages/pydantic/main.py", line 164, in __init__
    __pydantic_self__.__pydantic_validator__.validate_python(data, self_instance=__pydantic_self__)
pydantic_core._pydantic_core.ValidationError: 2 validation errors for InputPayload
dependencies
  Input should be a valid list [type=list_type, input_value=None, input_type=NoneType]
    For further information visit https://errors.pydantic.dev/2.5/v/list_type
inventory
  Value error, Inventory file "/tmp/dtt1-poc/manager/inventory.yaml" does not exist [type=value_error, input_value='/tmp/dtt1-poc/manager/inventory.yaml', input_type=str]
    For further information visit https://errors.pydantic.dev/2.5/v/value_error
.
[2024-02-15 18:55:03] [INFO] [465591] [MainThread] [workflow_engine]: Executing Reverse DAG tasks.
[2024-02-15 18:55:03] [INFO] [465591] [MainThread] [workflow_engine]: Executing tasks in parallel.
[2024-02-15 18:55:03] [INFO] [465591] [ThreadPoolExecutor-1_0] [workflow_engine]: [allocate-manager] Starting task.
[2024-02-15 18:55:10] [INFO] [465591] [ThreadPoolExecutor-1_0] [workflow_engine]: [allocate-manager] Finished task in 6.40 seconds.
$ 
```


Test number 57: :green_circle: 

```
tasks:
  # Generic manager test task
  - task: "run-manager-tests"
    description: "Run tests uninstall for the manager."
    do:
      this: process
      with:
        path: python3
        args:
          - testing/main.py
          - inventory: "{working-dir}/manager/inventory.yaml"
          - dependencies:
            - manager: "{working-dir}/manager-{manager-os}/inventory-not-exist.yaml"
          - tests: "upgrade"
          - component: "manager"
          - wazuh-version: "4.7.1"
          - wazuh-revision: "40709"
    depends-on:
      #- "provision-install-{manager}"
      - "provision-manager"

  # Unique manager provision task
  - task: "provision-manager"
    description: "Provision the manager."
    do:
      this: process
      with:
        path: python3
        args:
          - provision/main.py
          - inventory-manager: "{working-dir}/manager-{manager-os}/inventory.yaml"
          - install:
            - component: wazuh-manager
              type: package
    depends-on:
      - "allocate-manager"

  # Unique manager allocate task
  - task: "allocate-manager"
    description: "Allocate resources for the manager."
    do:
      this: process
      with:
        path: python3
        args:
          - allocation/main.py
          - action: create
          - provider: "{infra-provider}"
          - size: large
          - composite-name: "{manager-os}"
          - inventory-output: "{working-dir}/manager-{manager-os}/inventory.yaml"
          - track-output: "{working-dir}/manager-{manager-os}/track.yaml"
    cleanup:
      this: process
      with:
        path: python3
        args:
          - allocation/main.py
          - action: delete
          - track-output: "{working-dir}/manager-{manager-os}/track.yaml"

```

Obtained result:

```
$ python3 -m workflow_engine workflow_engine/examples/dtt1-tests.yaml 
[2024-02-15 18:57:32] [INFO] [467840] [MainThread] [workflow_engine]: Executing DAG tasks.
[2024-02-15 18:57:32] [INFO] [467840] [MainThread] [workflow_engine]: Executing tasks in parallel.
[2024-02-15 18:57:32] [INFO] [467840] [ThreadPoolExecutor-0_0] [workflow_engine]: [allocate-manager] Starting task.
[2024-02-15 18:58:16] [INFO] [467840] [ThreadPoolExecutor-0_0] [workflow_engine]: [allocate-manager] Finished task in 43.90 seconds.
[2024-02-15 18:58:16] [INFO] [467840] [ThreadPoolExecutor-0_0] [workflow_engine]: [provision-manager] Starting task.
[2024-02-15 19:00:17] [INFO] [467840] [ThreadPoolExecutor-0_0] [workflow_engine]: [provision-manager] Finished task in 121.80 seconds.
[2024-02-15 19:00:17] [INFO] [467840] [ThreadPoolExecutor-0_0] [workflow_engine]: [run-manager-tests] Starting task.
[2024-02-15 19:00:18] [ERROR] [467840] [ThreadPoolExecutor-0_0] [workflow_engine]: [run-manager-tests] Task failed with error: Error executing process task Traceback (most recent call last):
  File "/home/fcaffieri/repos/wazuh-qa/deployability/modules/testing/main.py", line 23, in <module>
    Tester.run(InputPayload(**vars(parse_arguments())))
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/fcaffieri/repos/wazuh-qa/.venv/lib/python3.11/site-packages/pydantic/main.py", line 164, in __init__
    __pydantic_self__.__pydantic_validator__.validate_python(data, self_instance=__pydantic_self__)
pydantic_core._pydantic_core.ValidationError: 1 validation error for InputPayload
inventory
  Value error, Inventory file "/tmp/dtt1-poc/manager/inventory.yaml" does not exist [type=value_error, input_value='/tmp/dtt1-poc/manager/inventory.yaml', input_type=str]
    For further information visit https://errors.pydantic.dev/2.5/v/value_error
.
[2024-02-15 19:00:18] [INFO] [467840] [MainThread] [workflow_engine]: Executing Reverse DAG tasks.
[2024-02-15 19:00:18] [INFO] [467840] [MainThread] [workflow_engine]: Executing tasks in parallel.
[2024-02-15 19:00:18] [INFO] [467840] [ThreadPoolExecutor-1_0] [workflow_engine]: [allocate-manager] Starting task.
[2024-02-15 19:00:24] [INFO] [467840] [ThreadPoolExecutor-1_0] [workflow_engine]: [allocate-manager] Finished task in 6.39 seconds.
$
```

Test number 56: :green_circle: 

```
tasks:
  # Generic manager test task
  - task: "run-manager-tests"
    description: "Run tests uninstall for the manager."
    do:
      this: process
      with:
        path: python3
        args:
          - testing/main.py
          - inventory: "{working-dir}/manager/inventory-not-exists.yaml"
          - dependencies:
            - manager: "{working-dir}/manager-{manager-os}/inventory.yaml"
          - tests: "upgrade"
          - component: "manager"
          - wazuh-version: "4.7.1"
          - wazuh-revision: "40709"
    depends-on:
      #- "provision-install-{manager}"
      - "provision-manager"

  # Unique manager provision task
  - task: "provision-manager"
    description: "Provision the manager."
    do:
      this: process
      with:
        path: python3
        args:
          - provision/main.py
          - inventory-manager: "{working-dir}/manager-{manager-os}/inventory.yaml"
          - install:
            - component: wazuh-manager
              type: package
    depends-on:
      - "allocate-manager"

  # Unique manager allocate task
  - task: "allocate-manager"
    description: "Allocate resources for the manager."
    do:
      this: process
      with:
        path: python3
        args:
          - allocation/main.py
          - action: create
          - provider: "{infra-provider}"
          - size: large
          - composite-name: "{manager-os}"
          - inventory-output: "{working-dir}/manager-{manager-os}/inventory.yaml"
          - track-output: "{working-dir}/manager-{manager-os}/track.yaml"
    cleanup:
      this: process
      with:
        path: python3
        args:
          - allocation/main.py
          - action: delete
          - track-output: "{working-dir}/manager-{manager-os}/track.yaml"

```

Obtained result:

```
$ python3 -m workflow_engine workflow_engine/examples/dtt1-tests.yaml 
[2024-02-15 19:02:09] [INFO] [470172] [MainThread] [workflow_engine]: Executing DAG tasks.
[2024-02-15 19:02:09] [INFO] [470172] [MainThread] [workflow_engine]: Executing tasks in parallel.
[2024-02-15 19:02:09] [INFO] [470172] [ThreadPoolExecutor-0_0] [workflow_engine]: [allocate-manager] Starting task.
[2024-02-15 19:02:52] [INFO] [470172] [ThreadPoolExecutor-0_0] [workflow_engine]: [allocate-manager] Finished task in 43.01 seconds.
[2024-02-15 19:02:52] [INFO] [470172] [ThreadPoolExecutor-0_0] [workflow_engine]: [provision-manager] Starting task.
[2024-02-15 19:05:20] [INFO] [470172] [ThreadPoolExecutor-0_0] [workflow_engine]: [provision-manager] Finished task in 148.39 seconds.
[2024-02-15 19:05:20] [INFO] [470172] [ThreadPoolExecutor-0_0] [workflow_engine]: [run-manager-tests] Starting task.
[2024-02-15 19:05:20] [ERROR] [470172] [ThreadPoolExecutor-0_0] [workflow_engine]: [run-manager-tests] Task failed with error: Error executing process task Traceback (most recent call last):
  File "/home/fcaffieri/repos/wazuh-qa/deployability/modules/testing/main.py", line 23, in <module>
    Tester.run(InputPayload(**vars(parse_arguments())))
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/fcaffieri/repos/wazuh-qa/.venv/lib/python3.11/site-packages/pydantic/main.py", line 164, in __init__
    __pydantic_self__.__pydantic_validator__.validate_python(data, self_instance=__pydantic_self__)
pydantic_core._pydantic_core.ValidationError: 1 validation error for InputPayload
inventory
  Value error, Inventory file "/tmp/dtt1-poc/manager/inventory-not-exists.yaml" does not exist [type=value_error, input_value='/tmp/dtt1-poc/manager/inventory-not-exists.yaml', input_type=str]
    For further information visit https://errors.pydantic.dev/2.5/v/value_error
.
[2024-02-15 19:05:20] [INFO] [470172] [MainThread] [workflow_engine]: Executing Reverse DAG tasks.
[2024-02-15 19:05:20] [INFO] [470172] [MainThread] [workflow_engine]: Executing tasks in parallel.
[2024-02-15 19:05:20] [INFO] [470172] [ThreadPoolExecutor-1_0] [workflow_engine]: [allocate-manager] Starting task.
[2024-02-15 19:05:27] [INFO] [470172] [ThreadPoolExecutor-1_0] [workflow_engine]: [allocate-manager] Finished task in 6.45 seconds.
$ 
```

Test number 55: :green_circle: 

```
tasks:
  # Generic manager test task
  - task: "run-manager-tests"
    description: "Run tests uninstall for the manager."
    do:
      this: process
      with:
        path: python3
        args:
          - main.py
          - inventory: "{working-dir}/manager/inventory.yaml"
          - dependencies:
            - manager: "{working-dir}/manager-{manager-os}/inventory.yaml"
          - tests: "upgrade"
          - component: "manager"
          - wazuh-version: "4.7.1"
          - wazuh-revision: "40709"
    depends-on:
      #- "provision-install-{manager}"
      - "provision-manager"

  # Unique manager provision task
  - task: "provision-manager"
    description: "Provision the manager."
    do:
      this: process
      with:
        path: python3
        args:
          - provision/main.py
          - inventory-manager: "{working-dir}/manager-{manager-os}/inventory.yaml"
          - install:
            - component: wazuh-manager
              type: package
    depends-on:
      - "allocate-manager"

  # Unique manager allocate task
  - task: "allocate-manager"
    description: "Allocate resources for the manager."
    do:
      this: process
      with:
        path: python3
        args:
          - allocation/main.py
          - action: create
          - provider: "{infra-provider}"
          - size: large
          - composite-name: "{manager-os}"
          - inventory-output: "{working-dir}/manager-{manager-os}/inventory.yaml"
          - track-output: "{working-dir}/manager-{manager-os}/track.yaml"
    cleanup:
      this: process
      with:
        path: python3
        args:
          - allocation/main.py
          - action: delete
          - track-output: "{working-dir}/manager-{manager-os}/track.yaml"

```

Obtained result:

```
$ python3 -m workflow_engine workflow_engine/examples/dtt1-tests.yaml 
[2024-02-15 19:09:42] [INFO] [474542] [MainThread] [workflow_engine]: Executing DAG tasks.
[2024-02-15 19:09:42] [INFO] [474542] [MainThread] [workflow_engine]: Executing tasks in parallel.
[2024-02-15 19:09:42] [INFO] [474542] [ThreadPoolExecutor-0_0] [workflow_engine]: [allocate-manager] Starting task.
[2024-02-15 19:10:26] [INFO] [474542] [ThreadPoolExecutor-0_0] [workflow_engine]: [allocate-manager] Finished task in 44.47 seconds.
[2024-02-15 19:10:26] [INFO] [474542] [ThreadPoolExecutor-0_0] [workflow_engine]: [provision-manager] Starting task.
[2024-02-15 19:12:50] [INFO] [474542] [ThreadPoolExecutor-0_0] [workflow_engine]: [provision-manager] Finished task in 144.19 seconds.
[2024-02-15 19:12:51] [INFO] [474542] [ThreadPoolExecutor-0_0] [workflow_engine]: [run-manager-tests] Starting task.
[2024-02-15 19:12:51] [ERROR] [474542] [ThreadPoolExecutor-0_0] [workflow_engine]: [run-manager-tests] Task failed with error: Error executing process task python3: can't open file '/home/fcaffieri/repos/wazuh-qa/deployability/modules/main.py': [Errno 2] No such file or directory
.
[2024-02-15 19:12:51] [INFO] [474542] [MainThread] [workflow_engine]: Executing Reverse DAG tasks.
[2024-02-15 19:12:51] [INFO] [474542] [MainThread] [workflow_engine]: Executing tasks in parallel.
[2024-02-15 19:12:51] [INFO] [474542] [ThreadPoolExecutor-1_0] [workflow_engine]: [allocate-manager] Starting task.
[2024-02-15 19:12:57] [INFO] [474542] [ThreadPoolExecutor-1_0] [workflow_engine]: [allocate-manager] Finished task in 6.40 seconds.
$
```

Test number 54: :green_circle: 

```
tasks:
  # Generic manager test task
  - task: "run-manager-tests"
    description: "Run tests uninstall for the manager."
    do:
      this: process
      with:
        path: python378
        args:
          - testing/main.py
          - inventory: "{working-dir}/manager/inventory.yaml"
          - dependencies:
            - manager: "{working-dir}/manager-{manager-os}/inventory.yaml"
          - tests: "upgrade"
          - component: "manager"
          - wazuh-version: "4.7.1"
          - wazuh-revision: "40709"
    depends-on:
      #- "provision-install-{manager}"
      - "provision-manager"

  # Unique manager provision task
  - task: "provision-manager"
    description: "Provision the manager."
    do:
      this: process
      with:
        path: python3
        args:
          - provision/main.py
          - inventory-manager: "{working-dir}/manager-{manager-os}/inventory.yaml"
          - install:
            - component: wazuh-manager
              type: package
    depends-on:
      - "allocate-manager"

  # Unique manager allocate task
  - task: "allocate-manager"
    description: "Allocate resources for the manager."
    do:
      this: process
      with:
        path: python3
        args:
          - allocation/main.py
          - action: create
          - provider: "{infra-provider}"
          - size: large
          - composite-name: "{manager-os}"
          - inventory-output: "{working-dir}/manager-{manager-os}/inventory.yaml"
          - track-output: "{working-dir}/manager-{manager-os}/track.yaml"
    cleanup:
      this: process
      with:
        path: python3
        args:
          - allocation/main.py
          - action: delete
          - track-output: "{working-dir}/manager-{manager-os}/track.yaml"

```

Obtained result:

```
$ python3 -m workflow_engine workflow_engine/examples/dtt1-tests.yaml 
[2024-02-15 19:13:58] [INFO] [476754] [MainThread] [workflow_engine]: Executing DAG tasks.
[2024-02-15 19:13:58] [INFO] [476754] [MainThread] [workflow_engine]: Executing tasks in parallel.
[2024-02-15 19:13:58] [INFO] [476754] [ThreadPoolExecutor-0_0] [workflow_engine]: [allocate-manager] Starting task.
[2024-02-15 19:14:39] [INFO] [476754] [ThreadPoolExecutor-0_0] [workflow_engine]: [allocate-manager] Finished task in 41.31 seconds.
[2024-02-15 19:14:39] [INFO] [476754] [ThreadPoolExecutor-0_0] [workflow_engine]: [provision-manager] Starting task.
[2024-02-15 19:16:40] [INFO] [476754] [ThreadPoolExecutor-0_0] [workflow_engine]: [provision-manager] Finished task in 121.17 seconds.
[2024-02-15 19:16:40] [INFO] [476754] [ThreadPoolExecutor-0_0] [workflow_engine]: [run-manager-tests] Starting task.
[2024-02-15 19:16:40] [ERROR] [476754] [ThreadPoolExecutor-0_0] [workflow_engine]: [run-manager-tests] Task failed with error: [Errno 2] No such file or directory: 'python378'.
[2024-02-15 19:16:40] [INFO] [476754] [MainThread] [workflow_engine]: Executing Reverse DAG tasks.
[2024-02-15 19:16:40] [INFO] [476754] [MainThread] [workflow_engine]: Executing tasks in parallel.
[2024-02-15 19:16:40] [INFO] [476754] [ThreadPoolExecutor-1_0] [workflow_engine]: [allocate-manager] Starting task.
[2024-02-15 19:16:47] [INFO] [476754] [ThreadPoolExecutor-1_0] [workflow_engine]: [allocate-manager] Finished task in 6.43 seconds.
$ 
```

</details>

**Comment by pro-akim** [comment]
## Update
### Modules test evidence

<details> <summary>track and inventory files</summary>
```
akim@akim-PC:/tmp/dtt1-poc$ cat agent-linux-ubuntu-22.04-amd64/inventory.yaml 
ansible_host: 192.168.57.3
ansible_port: 22
ansible_ssh_private_key_file: /tmp/wazuh-qa/VAGRANT-E8F37878-B306-40CA-A592-D5655203697D/instance_key
ansible_user: vagrant

akim@akim-PC:/tmp/dtt1-poc$ cat agent-linux-ubuntu-22.04-amd64/track.yaml 
identifier: VAGRANT-E8F37878-B306-40CA-A592-D5655203697D
instance_dir: /tmp/wazuh-qa/VAGRANT-E8F37878-B306-40CA-A592-D5655203697D
key_path: /tmp/wazuh-qa/VAGRANT-E8F37878-B306-40CA-A592-D5655203697D/instance_key
provider: vagrant

akim@akim-PC:/tmp/dtt1-poc$ cat manager-linux-ubuntu-22.04-amd64/inventory.yaml 
ansible_host: 192.168.57.2
ansible_port: 22
ansible_ssh_private_key_file: /tmp/wazuh-qa/VAGRANT-490839AA-3771-4C26-AD07-6D3DB2A0B9AF/instance_key
ansible_user: vagrant

akim@akim-PC:/tmp/dtt1-poc$ cat manager-linux-ubuntu-22.04-amd64/track.yaml 
identifier: VAGRANT-490839AA-3771-4C26-AD07-6D3DB2A0B9AF
instance_dir: /tmp/wazuh-qa/VAGRANT-490839AA-3771-4C26-AD07-6D3DB2A0B9AF
key_path: /tmp/wazuh-qa/VAGRANT-490839AA-3771-4C26-AD07-6D3DB2A0B9AF/instance_key
provider: vagrant


```


</details>

**Comment by pro-akim** [comment]
## Update :red_circle: 

![image](https://github.com/wazuh/wazuh-qa/assets/125690423/a322ea87-7f02-44b6-960c-361b748a1374)

Agents are not connected to the Manager.  It should be checked one more time.

**Comment by fcaffieri** [comment]
LGTM

**Comment by QU3B1M** [comment]
LGTM!
